flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java

```diff
@@ -301,35 +383,119 @@ void testServiceShutDownWithSynchronizedDriver() throws Exception {
                 Preconditions.checkNotNull(
                         testingLeaderElectionDriverFactory.getCurrentLeaderDriver());
 
-        final CheckedThread isLeaderThread =
-                new CheckedThread() {
-                    @Override
-                    public void go() {
-                        currentLeaderDriver.isLeader();
-                    }
-                };
-        isLeaderThread.start();
+        currentLeaderDriver.isLeader();
 
         leaderElectionService.stop();
-        isLeaderThread.sync();
+    }
+
+    @Test
+    void testOnLeadershipChangeDoesNotBlock() throws Exception {
+        new Context() {
+            {
+                runTestWithManuallyTriggeredEvents(
+                        executorService -> {
+                            final UUID expectedSessionID = UUID.randomUUID();
+
+                            testingLeaderElectionDriver.isLeader(expectedSessionID);
+                            executorService.trigger();
+
+                            testingContender.waitForLeader();
+
+                            // this call would time out if the
+                            // DefaultLeaderElectionService.onLeadershipChange would be called
+                            // synchronously
+                            testingLeaderElectionDriver.leaderInformationChanged(
+                                    LeaderInformation.empty());
+
+                            assertThat(testingLeaderElectionDriver.getLeaderInformation().isEmpty())
+                                    .isTrue();
+
+                            executorService.trigger();
+
+                            assertThat(
+                                            testingLeaderElectionDriver
+                                                    .getLeaderInformation()
+                                                    .getLeaderSessionID())
+                                    .isEqualTo(expectedSessionID);
+                        });
+            }
+        };
+    }
+
+    @Test
+    void testOnGrantLeadershipAsyncDoesNotBlock() throws Exception {
+        new Context() {
+            {
+                runTestWithManuallyTriggeredEvents(
+                        executorService -> {
+                            final UUID expectedSessionID = UUID.randomUUID();
+                            testingLeaderElectionDriver.isLeader(expectedSessionID);
+
+                            assertThat(leaderElectionService.getLeaderSessionID()).isNull();
+
+                            executorService.trigger();
+
+                            assertThat(leaderElectionService.getLeaderSessionID())
+                                    .isEqualTo(expectedSessionID);
+                        });
+            }
+        };
+    }
+
+    @Test
+    void testOnRevokeLeadershipDoesNotBlock() throws Exception {
```

Comment:
To be honest, I didn't understand the intention of this test. If we run it in a synchronized manner, it shouldn't timeout either.

Reply:
You're correct. Yesterday wasn't my strongest performance in that regards :innocent: I updated the tests.